### PR TITLE
Refactored top-k metrics and created TopKMetricsAggregator for optimized metrics computation

### LINKED ISSUE #491: [BUG] Performance regression after the refactory that moves losses and metrics to model.compile()
### Bug description
We observed a strong performance (throughput) regression after the refactory that moved losses and metrics to `model.compile()`.
The *average examples / sec* metric dropped from ~105k to 42k. That can be reproduced using the [retrieval experiments scripts](https://github.com/NVIDIA-Merlin/research/blob/main/retrieval_exp/scripts/retrieval_train_eval.py) with the same arguments, only moving the loss and metrics out of the task to the `model.compile()`.

After using some commits to isolate the issue as suggested by @marcromeyn , it seems the problem is introduced by some commit after this [commit [A]](https://github.com/NVIDIA-Merlin/models/commit/168ad03f8089146647c23b5fbe5fc65b39779ecb) and before (including) this [commit [B]](https://github.com/NVIDIA-Merlin/models/commit/10a3b8132c89991e7d25b0101c8e811654f24458).

Isolating a bit the issue, the lower throughput seems to be related to the metrics computation during training, as I also tested the pipeline (1) after removing all ranking metrics eval during training (Recall, NDCG, MRR) and (2) removing all metrics, including custom ones where I aggregate statistics from the logits.

In the following plot, it is possible to see the throughput at commit A (before the change) in **brown**, the throughput at commit B in **purple**, the throughput after removing ranking metrics (1) in **blue**, and the throughput after removing all but one metric (2) in **green**, which is closer to the original throughput (brown) 

image

So I suspect that some mechanisms might not be working properly after the refactory:
1 - The `train_metrics_steps` arg for `model.fit()` that configures training to compute metric values only each n steps is not working properly after the refactory.
2 Maybe that mechanism that sort predictions and targets only once for all ranking metrics and different cut-offs is not working anymore

More details on these perf tests in this [Slack thread (Nvidia Merlin internal)](https://nvidia.slack.com/archives/C020P24SQJJ/p1654152215076019)

### PR DESCRIPTION
**Pull request by gabrielspmoreira**
### Goals :soccer:
This PR improves metrics computation performance (both during training and inference) by sorting predictions only once before computing Top-k metrics. It is an addition improvement on top of #511
It also renames ranking metrics to top-k metrics which is more general (as RecallAt and PrecisionAt are top-k metrics and not ranking metrics).

### Implementation Details :construction:
- The `TopKMetricsAggregator` metric is introduced to aggregate a list of `TopkMetric` and sort predictions only once before computing the top-k metrics.
- I changed the top-k metrics `update_state(self,y_true,y_pred,sample_weight, label_relevant_counts)` to follow the standard Keras signature, by removing the `label_relevant_counts` arg, which cannot be passed to the standard `model.compiled_metrics.update_state()` method we are now using. To be able to pass that info to `TopkMetric` and to `TopKMetricsAggregator` a property`label_relevant_counts` was created for those classes, so that `model.compute_metrics()` can set it for all metrics before calling `update_state()`

### Testing Details :mag:
- Unit tests were updated to reflect the changes
- I ran a performance test on NGC without and with `TopKMetricsAggregator` 
1. Using `model.compile(..., metrics=[RecallAt(100), RecallAt(300), RecallAt(500), MRRAt(100), MRRAt(300), MRRAt(500), NDCGAt(100), NDCGAt(300), NDCGAt(500)])` 
2. Using  `model.compile(..., metrics=[TopKMetricsAggregator([RecallAt(100), RecallAt(300), ...])]`. 

You can see in the following plot that the training throughtput (average examples per second) increases from 40k to 55k when using `TopKMetricsAggregator`. 

image

P.s. This test was performed setting `model.fit(..., train_metrics_steps=1)` which computes metrics for each training step. The throughput of this experiment is improved to 97k (see #511) when we set `train_metrics_steps=100`.

### CODE REVIEW

File: merlin/models/tf/metrics/topk.py [hunk 1]
```diff
@@ -282,43 +297,126 @@ def from_config(cls, config):
...
-    metrics: List[RankingMetric] = []
+class TopKMetricsAggregator(Metric, TopkMetricWithLabelRelevantCountsMixin):
+    """Aggregator for top-k metrics (TopkMetric) that is optimized
+    to sort top-k predictions only once for all metrics.
+
+    topk_metrics : Sequence[TopkMetric]
+        Multiple arguments with TopkMetric instances
+    """
+
+    def __init__(self, topk_metrics: Sequence[TopkMetric]):
```
Comment (marcromeyn) [inline merlin/models/tf/metrics/topk.py:343]:
```suggestion
    def __init__(self, *topk_metrics: TopkMetric):
```

Comment (marcromeyn) [reply merlin/models/tf/metrics/topk.py:343]: This would simplify the API a bit. Since you could do: `TopKMetricsAggregator(RecallAt(5), MRRAt(5))`, instead of having to wrap it into a list.

Comment (gabrielspmoreira) [reply merlin/models/tf/metrics/topk.py]: Done

File: merlin/models/tf/metrics/topk.py [hunk 2]
```diff
@@ -282,43 +297,126 @@ def from_config(cls, config):
...
+
+    def result(self):
+        outputs = {}
+        for metric in self.topk_metrics:
+            outputs[metric.name] = metric.result()
+
+        return outputs
+
+
+def topk_metrics_default(top_ks: Sequence[int], **kwargs) -> Sequence[TopkMetric]:
```
[the remark this answers is not in the document]
Comment (gabrielspmoreira) [reply merlin/models/tf/metrics/topk.py]: Makes sense. Moved to a `TopKMetricsAggregator.default_metrics()` classmethod as you have suggested

**Review by marcromeyn — APPROVED** [review]
